libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1005,24 +1018,75 @@ string YulUtilFunctions::clearStorageRangeFunction(Type const& _type)
 {
 	string functionName = "clear_storage_range_" + _type.identifier();
 
-	solAssert(_type.storageBytes() >= 32, "Expected smaller value for storage bytes");
-
 	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
-			function <functionName>(start, end) {
+			function <functionName>(start, end, firstInactiveInStart) {
+				if gt(firstInactiveInStart, 0) {
+					let oldSlotValue := sload(start)
+					let mask := not(<shiftLeft>(mul(firstInactiveInStart, 8), <allOnes>))
+					sstore(start, and(oldSlotValue, mask))
+					start := add(start, 1)
+				}
 				for {} lt(start, end) { start := add(start, <increment>) }
 				{
-					<setToZero>(start, 0)
+					<setToZero>
 				}
 			}
 		)")
 		("functionName", functionName)
-		("setToZero", storageSetToZeroFunction(_type))
+		("allOnes", u256(-1).str())
+		("shiftLeft", shiftLeftFunctionDynamic())
+		(
+			"setToZero",
+			_type.storageBytes() < 17 ?
+			(storageSetWholeSlotToZeroFunction(_type) + "(start)") :
+			(storageSetToZeroFunction(_type) + "(start, 0)")
+		)
 		("increment", _type.storageSize().str())
 		.render();
 	});
 }
 
+string YulUtilFunctions::clearStorageStructFunction(StructType const& _type)
```

Comment:
I think it would probably be best to move this part to another PR as well.
You can implement ``delete`` for storage structs with this, that can again be done independently of the rest and can be tested on its own and then you can just reuse it in this PR.

Reply:
#9843